### PR TITLE
Guard Task execution via changed files

### PR DESCRIPTION
**Pull request by zregvart**
The new Task in `.tekton/tasks/changed-dirs.yaml` produces a list of changed directories for a particular pull request as an array result. Given this result guards can be done, e.g. using CEL expressions to limit when a Task on the Pipeline needs to run.
This way we can skip expensive Tasks that are unrelated to the change done in the pull request.

Similar to work in #1188 and #524, with the distinction that the PipelineRun is executed, only potentially not in full.

### CODE REVIEW

**Comment by zregvart** [comment]
Marked as draft, not sure if this approach will work, any further ideas welcome!

File: .tekton/tasks/changed-dirs.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        for path in $(gh pr view "https://github.com/konflux-ci/build-definitions/pull/$(params.pr_number)" --json files --jq '.files.[].path'); do
+          dirs+=("${path%%/*}")
+        done
+        # deduplicate dirs
+        readarray -t dirs < <(printf '%s\n' "${dirs[@]}" | sort -u)
+        # join $dirs with '","', e.g. (a b c) -> 'a","b","c","'
+        escaped_json="$(printf '%s\",\"' "${dirs[@]}")"
+        # add the square brackets and remove the trailing ',"'
+        escaped_json="[\"${escaped_json::-2}]"
+        echo -n "${escaped_json}" > "$(results.changed_directories.path)"
```
Comment (lcarva) [inline .tekton/tasks/changed-dirs.yaml:47]: The number of directories in this repo will blow past the result limit.

```
$ find . -type d | wc -c | numfmt --to=iec
11K
```

Comment (zregvart) [reply .tekton/tasks/changed-dirs.yaml:47]: It's only creating an array of top-level directories, so the current size is:

```
$ ls -d */|jq -cRn '[inputs]'|wc -c
107
```

Comment (lcarva) [reply .tekton/tasks/changed-dirs.yaml:47]: Great!

**Comment by chmeliik** [comment]
It does seem too limiting without `enable-cel-in-whenexpression` For many of the steps, guarding them on one specific directory is not enough :(

**Comment by zregvart** [comment]
> It does seem too limiting without `enable-cel-in-whenexpression` For many of the steps, guarding them on one specific directory is not enough :(

I agree, the alternative, as you pointed out is to not build a generic solution but a specific one, which would mean moving the logic into the `changed-dirs` Task (and in the process probably renaming it to something like `task-switchboard`). Tedious to maintain and not nearly as nice.

I do think the first iteration could be to look at the top level directories, that might be good enough.

There could be a middle ground, e.g. provide a set of expressions (in CEL or something else) and when those evaluate as `true` emit one item in the result array of `changed-dirs`, then use `when` based on that value. If that makes sense. I can try to prototype this idea on this PR.

> I was going to create a script that gets the list of changed files _and_ evaluates some gitignore-style conditions in one step to enable skipping e2e-tests. It would avoid the result size limit problems and be more flexible than the default `when` (but less flexible than `when` with CEL)

I think that kind of logic should go to the e2e harness, e.g. using [`Skip`](https://onsi.github.io/ginkgo/#skipping-specs) based on the Tasks/Pipelines changed. Might be easier to maintain this in golang, with a bit of effort this could payoff greatly.

**Comment by zregvart** [comment]
This version evaluates expressions over all files in a PR using OPA/Rego. The [`expressions` parameter](https://github.com/konflux-ci/build-definitions/pull/1558/files#diff-e1eb482b2093660ed8e19db808321a572bbfb86758055c532eadd30f039c0ddeR47-R52) determine by evaluating to `true` what result is omitted, the function [`strings.any_prefix_match`](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-strings-stringsany_prefix_match) comes in quite handy.

File: .tekton/pull-request.yaml [hunk 2]
```diff
@@ -36,7 +36,27 @@ spec:
...
+          - name: pr_number
+            value: "{{ pull_request_number }}"
+          - name: utils_image
+            value: quay.io/konflux-ci/pull-request-builds:appstudio-utils-{{revision}}
+          - name: expressions
+            value:
+              - tasks := strings.any_prefix_match(input, "task/")
+              - tasks_pipelines := strings.any_prefix_match(input, ["task/", "pipelines/"])
+              - e2e_tests := strings.any_prefix_match(input, ["task/", "pipelines/"])
+              - check_partner_tasks := strings.any_prefix_match(input, "partners/")
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply .tekton/pull-request.yaml:52]: (could also get more specific with it, but `.tekton/` and `hack/` would be good enough for me)

[the remark this answers is not in the document]
Comment (zregvart) [reply .tekton/pull-request.yaml]: Very true, added...

**Review by chmeliik — APPROVED** [review]
LGTM :taco:

**Pull request merged**